File: bindings/csharp/Satox/NFT/NFTManager.cs
using System;
using System.Runtime.InteropServices;
using satox_bindings;
using Satox.Bindings.NFT;

namespace Satox.NFT
{
    /// <summary>
    /// NFTManager provides NFT management functionality.
    /// </summary>
    public interface INFTManager : IDisposable
    {
        void Initialize();
        void Shutdown();
        NFT CreateNFT(string name, string description, string uri, string owner);
        NFT GetNFT(string nftId);
        NFT UpdateNFT(NFT nft);
        void DeleteNFT(string nftId);
        NFT[] GetAllNFTs();
        NFT[] GetNFTsByOwner(string owner);
        NFT TransferNFT(string nftId, string from, string to);
    }

    public class NFTManager : INFTManager
    {
        private readonly Satox.Bindings.NFT.NFTManager _manager;

        /// <summary>
        /// Creates a new instance of NFTManager.
        /// </summary>
        public NFTManager()
        {
            _manager = new Satox.Bindings.NFT.NFTManager();
        }

        /// <summary>
        /// Initializes the NFT manager.
        /// </summary>
        public void Initialize()
        {
            _manager.Initialize();
        }

        /// <summary>
        /// Shuts down the NFT manager.
        /// </summary>
        public void Shutdown()
        {
            _manager.Shutdown();
        }

        /// <summary>
        /// Creates a new NFT.
        /// </summary>
        /// <param name="name">The NFT name.</param>
        /// <param name="description">The NFT description.</param>
        /// <param name="uri">The NFT URI.</param>
        /// <param name="owner">The NFT owner.</param>
        /// <returns>The created NFT.</returns>
        public NFT CreateNFT(string name, string description, string uri, string owner)
        {
            return _manager.CreateNFT(name, description, uri, owner);
        }

        /// <summary>
        /// Gets an NFT by its ID.
        /// </summary>
        /// <param name="nftId">The NFT ID.</param>
        /// <returns>The NFT information.</returns>
        public NFT GetNFT(string nftId)
        {
            return _manager.GetNFT(nftId);
        }

        /// <summary>
        /// Updates an NFT.
        /// </summary>
        /// <param name="nft">The NFT to update.</param>
        /// <returns>The updated NFT.</returns>
        public NFT UpdateNFT(NFT nft)
        {
            return _manager.UpdateNFT(nft);
        }

        /// <summary>
        /// Deletes an NFT.
        /// </summary>
        /// <param name="nftId">The NFT ID.</param>
        public void DeleteNFT(string nftId)
        {
            _manager.DeleteNFT(nftId);
        }

        /// <summary>
        /// Gets all NFTs.
        /// </summary>
        /// <returns>An array of NFTs.</returns>
        public NFT[] GetAllNFTs()
        {
            return _manager.GetAllNFTs();
        }

        /// <summary>
        /// Gets NFTs by owner.
        /// </summary>
        /// <param name="owner">The owner address.</param>
        /// <returns>An array of NFTs.</returns>
        public NFT[] GetNFTsByOwner(string owner)
        {
            return _manager.GetNFTsByOwner(owner);
        }

        /// <summary>
        /// Transfers an NFT.
        /// </summary>
        /// <param name="nftId">The NFT ID.</param>
        /// <param name="from">The sender address.</param>
        /// <param name="to">The recipient address.</param>
        /// <returns>The updated NFT.</returns>
        public NFT TransferNFT(string nftId, string from, string to)
        {
            return _manager.TransferNFT(nftId, from, to);
        }

        public void Dispose()
        {
            _manager.Dispose();
        }
    }

    /// <summary>
    /// Represents an NFT.
    /// </summary>
    [StructLayout(LayoutKind.Sequential)]
    public struct NFT
    {
        /// <summary>
        /// The NFT ID.
        /// </summary>
        [MarshalAs(UnmanagedType.LPStr)]
        public string Id;

        /// <summary>
        /// The NFT name.
        /// </summary>
        [MarshalAs(UnmanagedType.LPStr)]
        public string Name;

        /// <summary>
        /// The NFT description.
        /// </summary>
        [MarshalAs(UnmanagedType.LPStr)]
        public string Description;

        /// <summary>
        /// The NFT URI.
        /// </summary>
        [MarshalAs(UnmanagedType.LPStr)]
        public string Uri;

        /// <summary>
        /// The NFT owner.
        /// </summary>
        [MarshalAs(UnmanagedType.LPStr)]
        public string Owner;

        /// <summary>
        /// The creation timestamp.
        /// </summary>
        public long CreatedAt;

        /// <summary>
        /// The last update timestamp.
        /// </summary>
        public long UpdatedAt;

        /// <summary>
        /// The number of transfers.
        /// </summary>
        public int TransferCount;
    }
} 

File: bindings/go/satox/nft/nft_manager.go
package nft

import (
	"encoding/json"
	"errors"
	"sync"
	"time"
)

// NFTMetadata represents the metadata for an NFT
type NFTMetadata struct {
	Name        string          `json:"name"`
	Description string          `json:"description"`
	Image       string          `json:"image"`
	ExternalURL string          `json:"externalUrl"`
	Attributes  json.RawMessage `json:"attributes"`
	Properties  json.RawMessage `json:"properties"`
}

// NFT represents a non-fungible token
type NFT struct {
	ID                 string      `json:"id"`
	ContractAddress    string      `json:"contractAddress"`
	Owner              string      `json:"owner"`
	Creator            string      `json:"creator"`
	Metadata           NFTMetadata `json:"metadata"`
	TokenID            uint64      `json:"tokenId"`
	TokenURI           string      `json:"tokenUri"`
	IsTransferable     bool        `json:"isTransferable"`
	IsBurnable         bool        `json:"isBurnable"`
	RoyaltyRecipient   string      `json:"royaltyRecipient"`
	RoyaltyBasisPoints uint32      `json:"royaltyBasisPoints"`
	CreatedAt          int64       `json:"createdAt"`
	UpdatedAt          int64       `json:"updatedAt"`
}

// NFTManager manages NFTs in the system
type NFTManager struct {
	nfts         map[string]*NFT
	ownerNFTs    map[string][]string // owner -> []nftID
	contractNFTs map[string][]string // contract -> []nftID
	creatorNFTs  map[string][]string // creator -> []nftID
	mutex        sync.RWMutex
	initialized  bool
}

// NewNFTManager creates a new NFTManager instance
func NewNFTManager() *NFTManager {
	return &NFTManager{
		nfts:         make(map[string]*NFT),
		ownerNFTs:    make(map[string][]string),
		contractNFTs: make(map[string][]string),
		creatorNFTs:  make(map[string][]string),
	}
}

// Initialize initializes the NFTManager
func (nm *NFTManager) Initialize() error {
	nm.mutex.Lock()
	defer nm.mutex.Unlock()

	if nm.initialized {
		return errors.New("NFTManager already initialized")
	}

	nm.initialized = true
	return nil
}

// CreateNFT creates a new NFT
func (nm *NFTManager) CreateNFT(contractAddress string, metadata NFTMetadata, creator string, isTransferable, isBurnable bool, royaltyRecipient string, royaltyBasisPoints uint32) (string, error) {
	nm.mutex.Lock()
	defer nm.mutex.Unlock()

	if !nm.initialized {
		return "", errors.New("NFTManager not initialized")
	}

	// Validate metadata
	if err := nm.validateMetadata(metadata); err != nil {
		return "", err
	}

	// Generate NFT ID
	nftID := generateNFTID(contractAddress, metadata)

	// Create NFT
	nft := &NFT{
		ID:                 nftID,
		ContractAddress:    contractAddress,
		Owner:              creator, // Creator is initial owner
		Creator:            creator,
		Metadata:           metadata,
		TokenID:            uint64(len(nm.nfts) + 1),
		TokenURI:           metadata.ExternalURL,
		IsTransferable:     isTransferable,
		IsBurnable:         isBurnable,
		RoyaltyRecipient:   royaltyRecipient,
		RoyaltyBasisPoints: royaltyBasisPoints,
		CreatedAt:          time.Now().Unix(),
		UpdatedAt:          time.Now().Unix(),
	}

	// Store NFT
	nm.nfts[nftID] = nft

	// Update indices
	nm.ownerNFTs[creator] = append(nm.ownerNFTs[creator], nftID)
	nm.contractNFTs[contractAddress] = append(nm.contractNFTs[contractAddress], nftID)
	nm.creatorNFTs[creator] = append(nm.creatorNFTs[creator], nftID)

	return nftID, nil
}

// GetNFT retrieves an NFT by ID
func (nm *NFTManager) GetNFT(nftID string) (*NFT, error) {
	nm.mutex.RLock()
	defer nm.mutex.RUnlock()

	if !nm.initialized {
		return nil, errors.New("NFTManager not initialized")
	}

	nft, exists := nm.nfts[nftID]
	if !exists {
		return nil, errors.New("NFT not found")
	}

	return nft, nil
}

// TransferNFT transfers an NFT from one owner to another
func (nm *NFTManager) TransferNFT(nftID, from, to string) error {
	nm.mutex.Lock()
	defer nm.mutex.Unlock()

	if !nm.initialized {
		return errors.New("NFTManager not initialized")
	}

	// Check if NFT exists
	nft, exists := nm.nfts[nftID]
	if !exists {
		return errors.New("NFT not found")
	}

	// Check if NFT is transferable
	if !nft.IsTransferable {
		return errors.New("NFT is not transferable")
	}

	// Check if sender is the owner
	if nft.Owner != from {
		return errors.New("Sender is not the owner")
	}

	// Update NFT
	nft.Owner = to
	nft.UpdatedAt = time.Now().Unix()

	// Update indices
	nm.removeFromOwnerNFTs(from, nftID)
	nm.ownerNFTs[to] = append(nm.ownerNFTs[to], nftID)

	return nil
}

// BurnNFT burns an NFT
func (nm *NFTManager) BurnNFT(nftID, owner string) error {
	nm.mutex.Lock()
	defer nm.mutex.Unlock()

	if !nm.initialized {
		return errors.New("NFTManager not initialized")
	}

	// Check if NFT exists
	nft, exists := nm.nfts[nftID]
	if !exists {
		return errors.New("NFT not found")
	}

	// Check if NFT is burnable
	if !nft.IsBurnable {
		return errors.New("NFT is not burnable")
	}

	// Check if sender is the owner
	if nft.Owner != owner {
		return errors.New("Sender is not the owner")
	}

	// Remove NFT from indices
	nm.removeFromOwnerNFTs(owner, nftID)
	nm.removeFromContractNFTs(nft.ContractAddress, nftID)
	nm.removeFromCreatorNFTs(nft.Creator, nftID)

	// Remove NFT
	delete(nm.nfts, nftID)

	return nil
}

// GetNFTsByOwner gets all NFTs owned by an address
func (nm *NFTManager) GetNFTsByOwner(owner string) ([]*NFT, error) {
	nm.mutex.RLock()
	defer nm.mutex.RUnlock()

	if !nm.initialized {
		return nil, errors.New("NFTManager not initialized")
	}

	nftIDs, exists := nm.ownerNFTs[owner]
	if !exists {
		return []*NFT{}, nil
	}

	nfts := make([]*NFT, 0, len(nftIDs))
	for _, nftID := range nftIDs {
		if nft, exists := nm.nfts[nftID]; exists {
			nfts = append(nfts, nft)
		}
	}

	return nfts, nil
}

// validateMetadata validates NFT metadata
func (nm *NFTManager) validateMetadata(metadata NFTMetadata) error {
	if metadata.Name == "" {
		return errors.New("NFT name cannot be empty")
	}

	if metadata.Description == "" {
		return errors.New("NFT description cannot be empty")
	}

	if metadata.Image == "" {
		return errors.New("NFT image cannot be empty")
	}

	return nil
}

// generateNFTID generates a unique NFT ID
func generateNFTID(contractAddress string, metadata NFTMetadata) string {
	// In a real implementation, this would use a more sophisticated ID generation
	// For now, we'll use a simple concatenation of metadata fields
	return contractAddress + "_" + metadata.Name
}

// Helper functions for managing indices
func (nm *NFTManager) removeFromOwnerNFTs(owner, nftID string) {
	nfts := nm.ownerNFTs[owner]
	for i, id := range nfts {
		if id == nftID {
			nm.ownerNFTs[owner] = append(nfts[:i], nfts[i+1:]...)
			break
		}
	}
}

func (nm *NFTManager) removeFromContractNFTs(contract, nftID string) {
	nfts := nm.contractNFTs[contract]
	for i, id := range nfts {
		if id == nftID {
			nm.contractNFTs[contract] = append(nfts[:i], nfts[i+1:]...)
			break
		}
	}
}

func (nm *NFTManager) removeFromCreatorNFTs(creator, nftID string) {
	nfts := nm.creatorNFTs[creator]
	for i, id := range nfts {
		if id == nftID {
			nm.creatorNFTs[creator] = append(nfts[:i], nfts[i+1:]...)
			break
		}
	}
}
